updates TextSelectCopyMulti to save content as file

flyinghermit opened the pull request:
Updates `TextSelectCopyMulti` component with download icon to save selected content as a file. 

<img width="1118" alt="Screenshot 2024-01-29 at 12 35 19 PM" src="https://github.com/gravitational/teleport/assets/10177602/e9a235ec-fa8c-45f8-a167-0a21850503c6">

flyinghermit committed:
feat: save content as file

flyinghermit commented on web/packages/teleport/src/components/TextSelectCopy/TextSelectCopyMulti.tsx:
```diff
@@ -55,12 +70,15 @@ export function TextSelectCopyMulti({ lines, bash = true }: Props) {
       pt={2}
       pr={7}
       borderRadius={2}
+      maxHeight={maxHeight}
```
`maxHeight` added to enable scroller for larger content. This is useful to display certificate content in SAML app configuration flow.

<img width="782" alt="Screenshot 2024-01-24 at 7 04 04 PM" src="https://github.com/gravitational/teleport/assets/10177602/f557b331-eb69-42ef-9b59-3687fdd33b4a">

zmb3 commented:
Is this needed for some particular task? Is there an issue you can link to?

flyinghermit commented:
> Is this needed for some particular task? Is there an issue you can link to?

There isn't a direct ticket to this one but the work stemmed from this parent issue [Add tctl command to print SAML IdP details](https://github.com/gravitational/teleport/issues/35564).

As shown here https://github.com/gravitational/teleport/pull/37232#discussion_r1465693262, the saml app configuration will now provide an option to copy or download IdP certificate. Since the content of the certificate will be shown and copyable with `TextSelectCopyMulti` component , I leaned towards adding a "download" icon button to this component, instead of of displaying another "download certificate" button.

flyinghermit committed:
add min height to have consistent height on empty text

flyinghermit committed:
run linter and update snapshot

flyinghermit committed:
Merge branch 'master' into sshah/textcopy-with-download

flyinghermit committed:
move download icon to the right of copy icon

flyinghermit commented:
Ping @kimlisa @ibeckermayer

kimlisa reviewed (commented):
can you add story with vertical scrolling too? (like in your sample image)

and a long text sample, because of:

<img width="694" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/7733119d-02d4-42e4-ab61-05e7701f6390">

flyinghermit committed:
increase left padding when showing download icon

flyinghermit committed:
accomodate vertical and horizontal scroller and icon button view

flyinghermit committed:
move download function to shared utils

flyinghermit commented on web/packages/teleport/src/components/TextSelectCopy/TextSelectCopyMulti.tsx:
```diff
@@ -44,6 +49,16 @@ export function TextSelectCopyMulti({ lines, bash = true }: Props) {
+    document.body.appendChild(a);
+    a.click();
+    document.body.removeChild(a);
+  }
```
moved to shared utils in https://github.com/gravitational/teleport/pull/37232/commits/a5f8383b21b3af0b396a819f0f2df24caae648fc

However, I haven't updated the http file download (the code you referenced). FYI, this can also replace the save function that is used in RDP terminal. But I'll have to test that once with proper file transfer and given my current bandwidth, I would prefer to follow up to that later on. Let me know if you still think that should be done in this PR. 

> also can this throw an error at anytime? (do we need like a try catch block?)

No try catch needed here as all is DOM operation. I tried to add it just in case but my editor complained `useless try catch block` :D

flyinghermit commented:
@kimlisa I have updated story as you suggested. I have also adjusted scroll view with respect to visible scroller so should address overlapping content as seen in your screenshot. https://github.com/gravitational/teleport/pull/37232/commits/2e78a371ec81f18435bd0e974b50a2f6d29ba2f3

flyinghermit committed:
rename downloadContentAsFile to saveOnDisk, unit test

flyinghermit committed:
original style with increased padding for download icon

flyinghermit committed:
Merge branch 'master' into sshah/textcopy-with-download

flyinghermit committed:
Merge branch 'master' into sshah/textcopy-with-download
